### ISSUE TITLE
clean_geometry test 2 hanging in Building::correct on MacOS

### ISSUE THREAD

**Issue by zeroset** [JuPedSim/jpscore#553]
Sometimes the geometry test case 2 is hanging on MacOS when correcting the geometry. 
The following Loop is never left again: 
```
for(auto wallPiece : WallPieces) {
    std::vector<Wall> tmpWallPieces;
    tmpWallPieces = SplitWall(subroom.second, wallPiece);
    if(!tmpWallPieces.empty()) {
        // std::cout << "set ok because tmp size =" << tmpWallPieces.size() << std::endl;
        ok = 0; /// stay in the loop
        // append tmpWallPieces to WallPiece
        WallPieces.insert(
            std::end(WallPieces),
            std::begin(tmpWallPieces),
            std::end(tmpWallPieces));
        // remove the line since it was split already
        auto it = std::find(WallPieces.begin(), WallPieces.end(), wallPiece);
        if(it != WallPieces.end()) {
            // std::cout<< KGRN << "delete wall ..." << RESET <<std::endl;
            // wallPiece.WriteToErrorLog();
            WallPieces.erase(it);
        }

        // std::cout << "BNow Wall peces size : " <<  WallPieces.size() << std::endl;
    }
}
```

It looks like one wall has a point with x coordinate almost +INF because it used to be something around `1.0 e154`

**Comment by zeroset** [comment]
@chraibi @schroedtert can you try to reproduce by starting jpscore with `systemtest/test_geometry/clean_geometry/inifile2.xml`

**Comment by schroedtert** [comment]
approx 1/5 cases it get stuck here

**Comment by Ozaq** [comment]
Not related to MacOS :)

I could reproduce it under ASAN:
```
==69035==ERROR: AddressSanitizer: heap-use-after-free on address 0x613000000ea8 at pc 0x000111495e12 bp 0x7ffee0f7c150 sp 0x7ffee0f7b900
READ of size 60 at 0x613000000ea8 thread T0
    #0 0x111495e11 in __asan_memcpy (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x5ee11)
    #1 0x10fa280a8 in Line::Line(Line const&) Line.h:42
    #2 0x10fa28396 in Wall::Wall(Wall const&) Wall.h:31
    #3 0x10fa0c60c in Wall::Wall(Wall const&) Wall.h:31
    #4 0x10f01dd32 in Building::correct() const Building.cpp:772
    #5 0x10f010149 in Building::InitGeometry() Building.cpp:289
    #6 0x10f00f512 in Building::Building(Configuration*, PedDistributor&) Building.cpp:72
    #7 0x10f013db4 in Building::Building(Configuration*, PedDistributor&) Building.cpp:63
    #8 0x10f9e31c8 in Simulation::InitArgs() Simulation.cpp:158
    #9 0x10ec88ecc in main main.cpp:87
    #10 0x7fff69df92e4 in start (libdyld.dylib:x86_64+0x112e4)

0x613000000ea8 is located 104 bytes inside of 384-byte region [0x613000000e40,0x613000000fc0)
freed by thread T0 here:
    #0 0x1114a5b02 in wrap__ZdlPv (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x6eb02)
    #1 0x10eca0a44 in std::__1::_DeallocateCaller::__do_call(void*) new:320
    #2 0x10eca09f8 in std::__1::_DeallocateCaller::__do_deallocate_handle_size(void*, unsigned long) new:278
    #3 0x10fa11cb4 in std::__1::_DeallocateCaller::__do_deallocate_handle_size_align(void*, unsigned long, unsigned long) new:254
    #4 0x10fa11c54 in std::__1::__libcpp_deallocate(void*, unsigned long, unsigned long) new:326
    #5 0x10fa80bdd in std::__1::allocator<Wall>::deallocate(Wall*, unsigned long) memory:1817
    #6 0x10fa80854 in std::__1::allocator_traits<std::__1::allocator<Wall> >::deallocate(std::__1::allocator<Wall>&, Wall*, unsigned long) memory:1555
    #7 0x10fa8a142 in std::__1::__split_buffer<Wall, std::__1::allocator<Wall>&>::~__split_buffer() __split_buffer:343
    #8 0x10fa896e4 in std::__1::__split_buffer<Wall, std::__1::allocator<Wall>&>::~__split_buffer() __split_buffer:340
    #9 0x10f0352ff in std::__1::enable_if<(__is_forward_iterator<std::__1::__wrap_iter<Wall*> >::value) && (is_constructible<Wall, std::__1::iterator_traits<std::__1::__wrap_iter<Wall*> >::reference>::value), std::__1::__wrap_iter<Wall*> >::type std::__1::vector<Wall, std::__1::allocator<Wall> >::insert<std::__1::__wrap_iter<Wall*> >(std::__1::__wrap_iter<Wall const*>, std::__1::__wrap_iter<Wall*>, std::__1::__wrap_iter<Wall*>) vector:2037
    #10 0x10f01e05a in Building::correct() const Building.cpp:779
    #11 0x10f010149 in Building::InitGeometry() Building.cpp:289
    #12 0x10f00f512 in Building::Building(Configuration*, PedDistributor&) Building.cpp:72
    #13 0x10f013db4 in Building::Building(Configuration*, PedDistributor&) Building.cpp:63
    #14 0x10f9e31c8 in Simulation::InitArgs() Simulation.cpp:158
    #15 0x10ec88ecc in main main.cpp:87
    #16 0x7fff69df92e4 in start (libdyld.dylib:x86_64+0x112e4)

previously allocated by thread T0 here:
    #0 0x1114a5502 in wrap__Znwm (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x6e502)
    #1 0x10fa1324c in std::__1::__libcpp_allocate(unsigned long, unsigned long) new:239
    #2 0x10fa84d21 in std::__1::allocator<Wall>::allocate(unsigned long, void const*) memory:1814
    #3 0x10fa84a60 in std::__1::allocator_traits<std::__1::allocator<Wall> >::allocate(std::__1::allocator<Wall>&, unsigned long) memory:1547
    #4 0x10fa89899 in std::__1::__split_buffer<Wall, std::__1::allocator<Wall>&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<Wall>&) __split_buffer:311
    #5 0x10fa8950c in std::__1::__split_buffer<Wall, std::__1::allocator<Wall>&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<Wall>&) __split_buffer:310
    #6 0x10fa88f61 in void std::__1::vector<Wall, std::__1::allocator<Wall> >::__push_back_slow_path<Wall const&>(Wall const&) vector:1618
    #7 0x10f02cb62 in std::__1::vector<Wall, std::__1::allocator<Wall> >::push_back(Wall const&) vector:1639
    #8 0x10f033344 in Building::SplitWall(std::__1::shared_ptr<SubRoom> const&, Wall const&) const Building.cpp:972
    #9 0x10f01d41b in Building::correct() const Building.cpp:760
    #10 0x10f010149 in Building::InitGeometry() Building.cpp:289
    #11 0x10f00f512 in Building::Building(Configuration*, PedDistributor&) Building.cpp:72
    #12 0x10f013db4 in Building::Building(Configuration*, PedDistributor&) Building.cpp:63
    #13 0x10f9e31c8 in Simulation::InitArgs() Simulation.cpp:158
    #14 0x10ec88ecc in main main.cpp:87
    #15 0x7fff69df92e4 in start (libdyld.dylib:x86_64+0x112e4)

SUMMARY: AddressSanitizer: heap-use-after-free (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x5ee11) in __asan_memcpy
Shadow bytes around the buggy address:
  0x1c2600000180: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
  0x1c2600000190: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c26000001a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c26000001b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c26000001c0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
=>0x1c26000001d0: fd fd fd fd fd[fd]fd fd fd fd fd fd fd fd fd fd
  0x1c26000001e0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c26000001f0: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
  0x1c2600000200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c2600000210: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c2600000220: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==69035==ABORTING
[1]    69035 abort      /Users/kkratz/fun/build-jpscore/bin/jpscore_asan
```

**Comment by Ozaq** [comment]
Just to be explicit: This does match the observation by @zeroset , due to the use-after-free the loop acts on memory that has been overwritten.

**Comment by Ozaq** [comment]
Analysis:
```
// Building.cpp:772
...
                    for(const auto& wallPiece : WallPieces) {
                        std::vector<Wall> tmpWallPieces;
                        tmpWallPieces = SplitWall(subroom.second, wallPiece);
                        if(!tmpWallPieces.empty()) {
                            // std::cout << "set ok because tmp size =" << tmpWallPieces.size() << std::endl;
                            ok = 0; /// stay in the loop
                            // append tmpWallPieces to WallPiece
                            WallPieces.insert(
                                std::end(WallPieces),
                                std::begin(tmpWallPieces),
                                std::end(tmpWallPieces));
                            // remove the line since it was split already
                            auto it = std::find(WallPieces.begin(), WallPieces.end(), wallPiece);
                            if(it != WallPieces.end()) {
                                // std::cout<< KGRN << "delete wall ..." << RESET <<std::endl;
                                // wallPiece.WriteToErrorLog();
                                WallPieces.erase(it);
                            }

                            // std::cout << "BNow Wall peces size : " <<  WallPieces.size() << std::endl;
                        }
                    }
...
```
So what I think is happening here is that: While iterating over the vector it is also appended. This is a problem because at some point the vector will reach max capacity and will be resized. This will invalidate the iterators used by the for loop. And then the loop variable actually points to freed memory.

**Comment by zeroset** [comment]
I am currently working on the fix, will open PR today/latest tomorrow.

**Comment by zeroset** [comment]
fixed in #592
